### ISSUE TITLE
Incorrect values not bound to statement

### ISSUE THREAD

**Issue by AlecStrong** [xerial/sqlite-jdbc#393]
```
java.sql.SQLException: Values not bound to statement
	at org.sqlite.core.CorePreparedStatement.checkParameters(CorePreparedStatement.java:71)
	at org.sqlite.jdbc3.JDBC3PreparedStatement.executeQuery(JDBC3PreparedStatement.java:76)
```

Occurs when I have SQL like:

```
SELECT *
FROM test
WHERE column = ?2
```

at runtime I should be able to `bindLong(2, someValue)` then execute the statement but I get the above exception because it thinks there are 2 total arguments.

Could just be expected behavior, but at least using the native C api the above works

**Comment by witmoca** [comment]
I don't think this has come up before. You are correct though. The native function `sqlite3_bind_parameter_count(sqlite3_stmt*)` of sqlite is used to get a count of how many parameters there are. Unfortunately according to the [sqlite reference](https://www.sqlite.org/c3ref/bind_parameter_count.html):

> This routine actually returns the index of the largest (rightmost) parameter. For all forms except ?NNN, this will correspond to the number of unique parameters.

I do recommend the obvious solution. Using the expected lower indexes certainly seems more logical.
If there is a (native) solution to get the correct parameter count, then we could support these types of bind parameters. It doesn't seem there is though.

**Comment by AlecStrong** [comment]
I'm on the library development side and not application development, so
downstream users might run into this problem and on my end I just have
tests failing. Would it make sense to just override the check somehow?
We're doing compile time checking of bind counts so the extra check in the
jdbc layer isn't really needed

**Comment by witmoca** [comment]
It's certainly possible. You could fork the jdbc and override the checks. The problem is that other implementations also rely on the bind_parameter_count being correct (and having parameters indexes that range from 1 to N). `Db#executeBatch(long stmt, int count, Object[] vals, boolean autoCommit)` uses it to bind the values for every comand in the batch for example. So you'd have to make sure that these still work. It's doable, but i'm not sure how viable a solution it is. 

In other words: overriding the checks is fairly easy, but the same issue is possibly going to cause other problems in sqlite-jdbc. You'd have to do a few in depth checks to see if it does.

**Comment by benasher44** [comment]
I just ran into this. I think this check can just be removed. Unbound parameters are technically allowed in sqlite3:

> … Unbound parameters are interpreted as NULL.

https://www.sqlite.org/c3ref/bind_blob.html

What are the dangers of removing the check? It looks like the count APIs are otherwise being used correctly, and removing the check should fix it. We'll likely fork internally and give it a go.

**Comment by benasher44** [comment]
Oh and also probably update the check here to not-bind for null values in the array: https://github.com/xerial/sqlite-jdbc/blob/ddd8fd3294cb3a8eea5f3e003702dbeb52d3327a/src/main/java/org/sqlite/core/DB.java#L793

**Comment by benasher44** [comment]
Okay so I've tested a fix locally that works. Basically the checkParameters() function and related code can be removed. I'm working on a PR with tests, and a broader explanation.
